### PR TITLE
Integrate Intel® Neural Compressor into Olive

### PR DESCRIPTION
**Pull request by yuwenzho**
Integrate Intel® Neural Compressor into Olive
new passes:
- IncStaticQuantization
- IncDynamicQuantization
- IncQuantization

### CODE REVIEW

File: olive/passes/onnx/quantization.py
Comment (ftian1) [inline olive/passes/onnx/quantization.py:412]: op_type_dict & op_name_dict should not be exposed to user

File: olive/passes/onnx/quantization.py
Comment (jambayk) [inline olive/passes/onnx/quantization.py:16]: Can we make `neural-compressor` an extra by doing a try import of these at `_run_for_config` line 552? Like in this openvino pass https://github.com/microsoft/Olive/blob/main/olive/passes/openvino/conversion.py#L57  

We could then add neural-compressor to extras here https://github.com/microsoft/Olive/blob/main/setup.py#L33 instead of making it a requirement.

File: examples/bert_inc_ptq_cpu/bert_inc_config.json
Comment (jambayk) [inline examples/bert_inc_ptq_cpu/bert_inc_config.json:62]: `"approach": "SEARCHABLE_VALUES"` is actually redundant since the pass has search enabled by default when the `search_strategy` is `None`. So, it will use the searchable_values when available.

**Comment by hshen14** [comment]
Hi @EmmaNingMS, we prepared the draft PR to integrate Intel Neural Compressor into Olive. Please review and feel free to add other folks to review. Your early comments are welcome. Thanks.

File: olive/passes/onnx/quantization.py
Comment (yuwenzho) [reply olive/passes/onnx/quantization.py]: Hi, I have moved `from neural_compressor import quantization, PostTrainingQuantConfig` into `_run_for_config`. Please help add neural-compressor to extras, Thanks!

Comment (jambayk) [reply olive/passes/onnx/quantization.py]: Thanks for moving it. Sorry, I was unclear in my comment. Can you please add the extra yourself to setup.py? Maybe you can call the extra as `inc`.

File: test/integ_test/evaluator/docker_eval/dockerfile/Dockerfile
Comment (jambayk) [inline test/integ_test/evaluator/docker_eval/dockerfile/Dockerfile:16]: I think these whitespaces will probably fail the formatting checks. 

Could you please install and run pre-commit? We have the hooks already defined. 
```
pip install -r requirements-dev.txt
pre-commit install 
# the first time, run on all files
pre-commit run --all-files
```

After the first time, pre-commit will automatically run on any changed files every time you git commit.

Comment (yuwenzho) [reply test/integ_test/evaluator/docker_eval/dockerfile/Dockerfile]: Thanks for your help. The PR code is now checked with pre-commit

File: olive/passes/onnx/quantization.py
Comment (yuwenzho) [reply olive/passes/onnx/quantization.py]: Sorry I misunderstood you. I have added the extra as `"inc": ["neural-compressor"],`

File: docs/source/tutorials/passes/onnx.md
Comment (yuwenzho) [reply docs/source/tutorials/passes/onnx.md]: Fixed. Thanks for your review.

File: olive/passes/onnx/quantization.py
Comment (yuwenzho) [reply olive/passes/onnx/quantization.py]: Fixed. Thanks for your review.

File: olive/passes/onnx/quantization.py
Comment (yuwenzho) [reply olive/passes/onnx/quantization.py:558]: Hi, thanks for your review. I added `quant_level` in config, which is the strategy parameter. INC optimized the strategy, and now the default value is 'auto'

**Comment by xiaoyu-work** [comment]
Can you add some unit tests for the new adding passes?

**Comment by yuwenzho** [comment]
> Can you add some unit tests for the new adding passes?

Hi, thanks for your review. I added UT test in the same way as `OnnxDynamicQuantization`.

**Comment by yuwenzho** [comment]
@jambayk @trajepl Hi, I noticed that some expected tests are not being started. Could you please provide some help?

File: test/unit_test/engine/test_engine.py
Comment (jambayk) [inline test/unit_test/engine/test_engine.py:78]: Hi, instead of these additional tests, Xiaoyu's comment meant some tests like those added in this PR https://github.com/microsoft/Olive/pull/184. 

The current changes to the tests in this PR are not required since these are tests for the engine.

File: olive/extra_dependencies.json
Comment (jambayk) [inline olive/extra_dependencies.json:24]: Could you also add references to this extra in this file https://github.com/microsoft/Olive/blob/main/olive/workflows/run/run.py#L59? Sorry, this was a recent change in the repo. Some of the code is still frequently changing so there are new breaking changes.

**Comment by jambayk** [comment]
> @jambayk @trajepl Hi, I noticed that some expected tests are not being started. Could you please provide some help?

Yes. Since this PR is from a forked repo, it needs a team member to add a comment that triggers the ci pipeline. I can do it once my comments above are resolved. Thanks!

File: test/unit_test/engine/test_engine.py
Comment (yuwenzho) [reply test/unit_test/engine/test_engine.py]: Sorry, I misunderstood. Please review the new unit tests.

File: olive/extra_dependencies.json
Comment (yuwenzho) [reply olive/extra_dependencies.json:24]: Thanks for your review. I have added references into `run.py`

File: test/requirements-test.txt
Comment (jambayk) [inline test/requirements-test.txt:12]: nit: can you add this in ascending order?

Comment (yuwenzho) [reply test/requirements-test.txt]: Fixed. Thanks for your review.

**Comment by yuwenzho** [comment]
@jambayk Hi, I just fixed some typo. Please help start the tests. Thanks!

**Comment by jambayk** [comment]
@yuwenzho The inc unit test is hanging on windows because the `neural_compressor` related imports are hanging for some reason on the ci pipeline agent. It works fine on my local windows machine. 
Could you cherry pick this commit or copy the changes in this commit 40c3e194d19f96a6b43ae4d93a1a69102abd53f4? I think we can skip the test on windows for now since it's running fine on Linux.

**Comment by yuwenzho** [comment]
@jambayk Hi, I updated inc example. Please help start the tests. Thanks!

**Comment by hshen14** [comment]
Hi @jambayk  @EmmaNingMS Please let us know whether the PR is good to be merged. Thanks.

**Comment by jambayk** [comment]
> Hi @jambayk @EmmaNingMS Please let us know whether the PR is good to be merged. Thanks.

@hshen14 It looks good to me. I have approved the PR. Thanks for your contribution!

**Review by jambayk — APPROVED** [review]

**Comment by yuwenzho** [comment]
@jambayk @EmmaNingMS Hi, would it be possible to merge this PR if there are no further comments at this time?

**Comment by devang-ml** [comment]
> @jambayk @EmmaNingMS Hi, would it be possible to merge this PR if there are no further comments at this time?

We need to enable CLA bot from our side before we can merge this PR.

**Comment by jambayk** [comment]
@yuwenzho Sorry for the delay. I resolved the merge conflicts but there are some test failures due to some updates we made recently in the main branch. I have added the fixes in this commit https://github.com/microsoft/Olive/commit/e4e8ef657efb2ab168de595f068dcb6f3cbb88ed. Could you please cherry pick the commit into this PR branch. After that, I thin we should be able to merge this PR.

**Review by jambayk — APPROVED** [review]

**Comment by jambayk** [comment]
Merged the PR. Thanks for your contribution! the PR now. Thanks for your contribution!

**Pull request merged**
